keps/sig-cli/3659-kubectl-apply-prune/README.md

```diff
@@ -0,0 +1,1381 @@
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [ ] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name:
+  - Components depending on the feature gate:
+- [X] Other
+  - Describe the mechanism:
+
+     Functionality is client-side only.  Plan for alpha is that users must explicitly
+     opt-in with `KUBECTL_APPLYSET_ALPHA`, existing functionality will remain.
+
+  - Will enabling / disabling the feature require downtime of the control
+    plane?
+
+      No
+
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+      No
+
+###### Does enabling the feature change any default behavior?
+
+<!--
+Any change of default behavior may be surprising to users or break existing
+automations, so be extremely careful here.
+-->
+
+If the user opts-in, then they will get the new functionality.  For alpha,
+the existing functionality will remain the default for alpha.  We do not
+plan to replace the existing functionality, users will always need to
+opt-in by specifying an applyset-id flag.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+<!--
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
+
+Feature gates are typically disabled by setting the flag to `false` and
+restarting the component. No other changes should be necessary to disable the
+feature.
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+Yes, this is client-side only.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Mixing and matching current pruning with new pruning and non-pruning
+might cause objects to not be pruned, as with today when mixing pruning
+modes.
+
+###### Are there any tests for feature enablement/disablement?
+
+<!--
+The e2e framework does not currently support enabling or disabling feature
+gates. However, unit tests in each component dealing with managing data, created
+with and without the feature, are necessary. At the very least, think about
+conversion tests if API types are being modified.
+
+Additionally, for features that are introducing a new API field, unit tests that
+are exercising the `switch` of feature gate itself (what happens if I disable a
+feature gate after having objects written with the new field) are also critical.
+You can take a look at one potential example of such test in:
+https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
+-->
+
+We will maintain tests for "prune v2" and "prune v1", until
+such time as prune v1 (technically still in alpha) is removed.
```

Comment:
Do you plan to add unit tests that will exercise the environment variable flag?

Reply:
Good call - yes - we can add to the doc.  We'll add unit tests (or cases) that cover v2 apply/prune, and leave the existing v1 apply/prune unit tests basically as-is until that functionality is removed.  I don't think there should be any possible "crosstalk" between the two, but if there is we'll make sure there's test coverage to make sure we haven't regressed.